mm/map/mm_map.c

```diff
@@ -0,0 +1,337 @@
+
+static bool vaddr_in_area(FAR const void *addr, FAR const void *start,
+                          size_t length)
+{
+  uintptr_t u_addr = (uintptr_t)addr;
+  uintptr_t u_start = (uintptr_t)start;
+  return (u_addr >= u_start) && (u_addr < u_start + length);
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: mm_map_initialize
+ *
+ * Description:
+ *   Allocates a task group specific mm_map stucture. Called when the group
+ *   is initialized
+ *
+ ****************************************************************************/
+
+void mm_map_initialize(FAR struct mm_map_s *mm)
+{
+  memset(mm, 0, sizeof(struct mm_map_s));
+  sq_init(&mm->mm_map_sq);
+  nxmutex_init(&mm->mm_map_mutex);
+}
+
+/****************************************************************************
+ * Name: mm_map_destroy
+ *
+ * Description:
+ *   De-allocates a task group specific mm_map stucture and the mm_map_mutex
+ *
+ ****************************************************************************/
+
+void mm_map_destroy(FAR struct mm_map_s *mm)
+{
+  FAR struct mm_map_entry_s *map;
+
+  while ((map = (FAR struct mm_map_entry_s *)sq_remfirst(&mm->mm_map_sq)))
+    {
+      /* Pass null as group argument to indicate that actual MMU mappings
+       * must not be touched. The process is being deleted and we don't
+       * know in which context we are. Only kernel memory allocations
+       * need to be freed by drivers
+       */
+
+      if (map->munmap)
+        {
+          /* Unmap the whole region */
+
+          if (map->munmap(NULL, map, (void *)map->vaddr, map->length) < 0)
+            {
+              /* This is a bug. A driver provides munmap, but can't munmap
+               * the region. Just delete the map entry.
+               */
+
+              kmm_free(map);
+            }
+        }
+      else
+        {
+          /* Mappping has been added, but there is no unmap. Just delete the
+           * map entry.
+           */
+
+          kmm_free(map);
+        }
+    }
+
+  nxmutex_destroy(&mm->mm_map_mutex);
+}
+
+/****************************************************************************
+ * Name: mm_map_add
+ *
+ * Description:
+ *   Add a mapping to task group's mm_map list
+ *
+ ****************************************************************************/
+
+int mm_map_add(FAR struct mm_map_entry_s *entry)
+{
+  FAR struct tcb_s *tcb = nxsched_self();
+  FAR struct task_group_s *group = tcb->group;
+  FAR struct mm_map_s *mm = &group->tg_mm_map;
+  FAR struct mm_map_entry_s *map = kmm_zalloc(sizeof(struct mm_map_entry_s));
+  int ret;
+
+  if (!entry)
+    {
+      return -EINVAL;
+    }
+
+  if (!map)
+    {
+      return -ENOMEM;
+    }
+
+  /* Copy the provided mapping and add to the list */
+
+  *map = *entry;
```

Comment:
why not?

1. kmm_malloc entry in the first place
2. or kmm_malloc here and return

Reply:
done cleanup in here